### PR TITLE
Using math::SpeedLimiter on the diff_drive controller.

### PR DESCRIPTION
**Pull request by LolaSegura**
Signed-off-by: LolaSegura <lsegura@ekumenlabs.com>

## Summary
This replace the diff_drive controller speed limiter class for the one implemented in ignition math. Part of #810. 

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [x] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸

### CODE REVIEW

File: src/systems/diff_drive/DiffDrive.cc [hunk 1]
```diff
@@ -197,56 +196,48 @@ void DiffDrive::Configure(const Entity &_entity,
...
+  // Instantiate the speed limiters.
+  this->dataPtr->limiterLin = std::make_unique<ignition::math::SpeedLimiter>();
+  this->dataPtr->limiterAng = std::make_unique<ignition::math::SpeedLimiter>();
 
+  // Parse speed limiter parameters.
   if (_sdf->HasElement("min_velocity"))
   {
-    minVel = _sdf->Get<double>("min_velocity");
-    hasVelocityLimits = true;
+    double minVel = _sdf->Get<double>("min_velocity");
```
Comment (francocipollone) [inline src/systems/diff_drive/DiffDrive.cc:206]: nit: it can be `const`

File: src/systems/diff_drive/DiffDrive.cc [hunk 2]
```diff
@@ -197,56 +196,48 @@ void DiffDrive::Configure(const Entity &_entity,
...
+    this->dataPtr->limiterLin->SetMinVelocity(minVel);
+    this->dataPtr->limiterAng->SetMinVelocity(minVel);
   }
   if (_sdf->HasElement("max_velocity"))
   {
-    maxVel = _sdf->Get<double>("max_velocity");
-    hasVelocityLimits = true;
+    double maxVel = _sdf->Get<double>("max_velocity");
```
Comment (francocipollone) [inline src/systems/diff_drive/DiffDrive.cc:212]: nit: it can be `const`

File: src/systems/diff_drive/DiffDrive.cc [hunk 3]
```diff
@@ -197,56 +196,48 @@ void DiffDrive::Configure(const Entity &_entity,
...
+    this->dataPtr->limiterLin->SetMaxVelocity(maxVel);
+    this->dataPtr->limiterAng->SetMaxVelocity(maxVel);
   }
   if (_sdf->HasElement("min_acceleration"))
   {
-    minAccel = _sdf->Get<double>("min_acceleration");
-    hasAccelerationLimits = true;
+    double minAccel = _sdf->Get<double>("min_acceleration");
```
Comment (francocipollone) [inline src/systems/diff_drive/DiffDrive.cc:218]: nit: it can be `const`

File: src/systems/diff_drive/DiffDrive.cc [hunk 4]
```diff
@@ -197,56 +196,48 @@ void DiffDrive::Configure(const Entity &_entity,
...
+    this->dataPtr->limiterLin->SetMinAcceleration(minAccel);
+    this->dataPtr->limiterAng->SetMinAcceleration(minAccel);
   }
   if (_sdf->HasElement("max_acceleration"))
   {
-    maxAccel = _sdf->Get<double>("max_acceleration");
-    hasAccelerationLimits = true;
+    double maxAccel = _sdf->Get<double>("max_acceleration");
```
Comment (francocipollone) [inline src/systems/diff_drive/DiffDrive.cc:224]: nit: it can be `const`

File: src/systems/diff_drive/DiffDrive.cc [hunk 5]
```diff
@@ -197,56 +196,48 @@ void DiffDrive::Configure(const Entity &_entity,
...
+    this->dataPtr->limiterLin->SetMaxAcceleration(maxAccel);
+    this->dataPtr->limiterAng->SetMaxAcceleration(maxAccel);
   }
   if (_sdf->HasElement("min_jerk"))
   {
-    minJerk = _sdf->Get<double>("min_jerk");
-    hasJerkLimits = true;
+    double minJerk = _sdf->Get<double>("min_jerk");
```
Comment (francocipollone) [inline src/systems/diff_drive/DiffDrive.cc:230]: nit: it can be `const`

File: src/systems/diff_drive/DiffDrive.cc [hunk 6]
```diff
@@ -197,56 +196,48 @@ void DiffDrive::Configure(const Entity &_entity,
...
+    this->dataPtr->limiterLin->SetMinJerk(minJerk);
+    this->dataPtr->limiterAng->SetMinJerk(minJerk);
   }
   if (_sdf->HasElement("max_jerk"))
   {
-    maxJerk = _sdf->Get<double>("max_jerk");
-    hasJerkLimits = true;
+    double maxJerk = _sdf->Get<double>("max_jerk");
```
Comment (francocipollone) [inline src/systems/diff_drive/DiffDrive.cc:236]: nit: it can be `const`

**Review by francocipollone — COMMENTED** [review]
LGTM! I just left a couple of nits.

I will add @scpeters as a reviewer as well

**Comment by scpeters** [comment]
this is looking good!

it's helpful to cross-reference issues from pull requests, so I would add that this is "Part of https://github.com/ignitionrobotics/ign-gazebo/issues/810" in the pull request description

File: src/systems/diff_drive/DiffDrive.cc (hunk 6 above)
Comment (LolaSegura) [reply src/systems/diff_drive/DiffDrive.cc]: Addressed all.

**Comment by scpeters** [comment]
I noticed that codecov is complaining about test coverage. We are not strict about requiring code coverage, but I think we can easily increase this for this test. I believe the `diff_drive` system has the following integration test:

* https://github.com/ignitionrobotics/ign-gazebo/blob/ign-gazebo3/test/integration/diff_drive_system.cc

The test uses several different world files (search for `test/worlds` to find [lines like this](https://github.com/ignitionrobotics/ign-gazebo/blob/ignition-gazebo3_3.8.0/test/integration/diff_drive_system.cc#L203)) such as [test/worlds/diff_drive.sdf](https://github.com/ignitionrobotics/ign-gazebo/blob/ignition-gazebo3_3.8.0/test/worlds/diff_drive.sdf). This world has a plugin that [specifies max_acceleration and max_velocity](https://github.com/ignitionrobotics/ign-gazebo/blob/ignition-gazebo3_3.8.0/test/worlds/diff_drive.sdf#L230-L231), but doesn't specify `min_acceleration` or `min_velocity`. If we added `min_acceleration` of `-1` and `min_velocity` of `0.5`, that will improve the coverage and add symmetry to the test. To get even more coverage and without changing the dynamics of the system we could add a `min_jerk` value of `-Inf` and a `max_jerk` value of `Inf` (I think that will work).

Since these are easy changes to improve code coverage, do you mind giving them a try? Thanks!

**Comment by LolaSegura** [comment]
> If we added `min_acceleration` of -1 and `min_velocity` of 0.5, that will improve the coverage and add symmetry to the test.

@scpeters  here when you mentioned the `min_velocity` did you meant to write -0.5? Because otherwise it would take the same value as the `max_velocity`.

**Comment by scpeters** [comment]
> > If we added `min_acceleration` of -1 and `min_velocity` of 0.5, that will improve the coverage and add symmetry to the test.
> 
> @scpeters here when you mentioned the `min_velocity` did you meant to write -0.5? Because otherwise it would take the same value as the `max_velocity`.

thanks for catching that; I meant to write `-0.5`

**Review by francocipollone — CHANGES_REQUESTED** [review]
ptal at the comments. Also, run `codecheck` to match style.

**Comment by francocipollone** [comment]
@LolaSegura . I've tried it locally.
 - It seems that just adding the `min_jerk` and `max_jerk` properties are causing a miss behavior in the diff drive when moving forward(probably backward too).
 - Leaving those properties commented out the test moving forward works fine.
 - It is failing only one test when moving backward:
 ```
 Expected: (expectedLowerPosition.Y()) < (expectedGreaterPosition.Y()), actual: 0.710173 vs 1.40412e-12
 ```
 Re-check the desire angle you are publishing.

**Comment by LolaSegura** [comment]
>If we added min_acceleration of -1 and min_velocity of 0.5 -0.5, that will improve the coverage and add symmetry to the test

I added the `<min_velocity>` and `<min_acceleration>` to the `diff_drive.sdf` and the `diff_drive_custom_topics.sdf`. With this the test its working fine. 

> To get even more coverage and without changing the dynamics of the system we could add a min_jerk value of -Inf and a max_jerk value of Inf (I think that will work).

I try adding this limits as well, but I encounter a problem because when being parsed the limits where interpreted as zero, causing the test to fail. An option would be setting the `<max_jerk>` with a big value and the `<min_jerk>` with a small one to get more test coverage.

**Comment by francocipollone** [comment]
It is ready for review @scpeters

**Review by scpeters — COMMENTED** [review]
> > To get even more coverage and without changing the dynamics of the system we could add a min_jerk value of -Inf and a max_jerk value of Inf (I think that will work).
> 
> I try adding this limits as well, but I encounter a problem because when being parsed the limits where interpreted as zero, causing the test to fail. An option would be setting the `<max_jerk>` with a big value and the `<min_jerk>` with a small one to get more test coverage.

I'll look into this briefly

File: test/integration/diff_drive_system.cc [hunk 7]
```diff
@@ -54,145 +54,170 @@ class DiffDriveTest : public ::testing::TestWithParam<int>
...
+
+      if(forward)
+      {
+        EXPECT_LT(v, 0.5);
+        EXPECT_LT(a, 1);
+      }
+      else
+      {
+        EXPECT_GT(v, -0.5);
+        EXPECT_GT(a, -1);
```
[the remark this answers is not in the document]
Comment (LolaSegura) [reply test/integration/diff_drive_system.cc]: Addressed both comments.

**Comment by scpeters** [comment]
> > To get even more coverage and without changing the dynamics of the system we could add a min_jerk value of -Inf and a max_jerk value of Inf (I think that will work).
> 
> I try adding this limits as well, but I encounter a problem because when being parsed the limits where interpreted as zero, causing the test to fail. An option would be setting the `<max_jerk>` with a big value and the `<min_jerk>` with a small one to get more test coverage.

thanks for testing this; I just looked into it a bit deeper and found that it is a parsing issue with `libsdformat` (https://github.com/osrf/sdformat/issues/261). We can leave out the jerk tests for now

**Review by scpeters — APPROVED** [review]

**Pull request merged**